models/User.js

```diff
@@ -0,0 +1,71 @@
+const mongoose = require('mongoose');
+const bcrypt = require('bcrypt');
+
+const saltRounds = 10;
+
+const Schema = mongoose.Schema;
+
+const UserSchema = new Schema({
+  userName: { type: String, required: true, unique: true },
+  passwordHash: { type: String, required: true },
+  conversation: { type: Schema.Types.ObjectId, ref: 'Conversation' },
+});
+
+
+UserSchema.set('toJSON', {
+  transform: (doc, ret) => {
+    const user = ret;
+    delete user.passwordHash;
+    return user;
+  },
+});
+
+UserSchema.methods.comparePassword = function (password, cb) {
+  bcrypt.compare(password, this.passwordHash, (err, isMatch) => {
+    if (err) return cb(err);
+
+    cb(null, isMatch);
+  });
+};
+
+UserSchema.statics.authenticate = function (userName, password, cb) {
+  this.findOne({ userName }, (err, user) => {
+    if (err) return cb(err);
+
+    if (!user) return cb({ error: 'user not found' });
+
+    user.comparePassword(password, (err, isMatch) => {
+      if (err) return cb(err);
+
+      if (isMatch) {
+        return cb(null, user);
+      }
+      return cb({ error: 'invalid password' });
```

Comment:
Since this module doesn't know about json error responses, the callback should return a plain javascript `Error`:
`cb(new Error('invalid password'))`